lib/ansible/modules/cloud/amazon/aws_s3_bucket_facts.py

```diff
@@ -0,0 +1,172 @@
+
+# Lists all s3 buckets
+- aws_s3_bucket_facts:
+'''
+
+RETURN = '''
+Buckets:
+  description: List of buckets
+  returned: always
+  type: list
+  sample:
+    [{
+        "Name": "my_bucket",
+        "CreationDate": "2017-07-06T15:05:12+00:00"
+    }]
+Owner:
+  description: Details of bucket owner
+  returned: always
+  type: dict
+  sample:
+    {
+      "DisplayName": "my_user",
+      "ID": "d5a0152fc8eb469fcc7eed35b2da939fd9e1a9c53924b681450f5353199900f8"
+    }
+ResponseMetadata:
+  description: Details of API call
+  returned: always
+  type: dict
+  sample:
+    {
+      "HTTPHeaders": {
+        "content-type": "application/xml",
+        "date": "Thu, 27 Jul 2017 11:37:07 GMT",
+        "server": "AmazonS3",
+        "transfer-encoding": "chunked",
+        "x-amz-id-2": "LeNLtFwr0DmQD7+pb52w3A+vuR9vkVNDswfJyTHj6673yBDoYS24EShCztue+ARrxqzDGsNWmOk=",
+        "x-amz-request-id": "F47C7B7622231E44"
+      },
+      "HTTPStatusCode": 200,
+      "HostId": "LeNLtFwr0DmQD7+pb52w3A+vuR9vkVNDswfJyTHj6673yBDoYS24EShCztue+ARrxqzDGsNWmOk=",
+      "RequestId": "F47C7B7622231E44",
+      "RetryAttempts": 0
+    }
+'''
+
+import json
+import traceback
+from datetime import datetime
+from json import dumps
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import get_aws_connection_info, ec2_argument_spec, boto3_conn, \
+    camel_dict_to_snake_dict
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError, NoCredentialsError
+
+    HAS_BOTO3 = True
+except ImportError:
+    HAS_BOTO3 = False
+
+
+class S3ServiceManager(object):
+    """
+    Service manager for S3
+    """
+
+    def __init__(self, module):
+        """
+        Bootstrap module and connection
+        :param module:
+        """
+        self.module = module
+
+        # Set up connection
+        region, ec2_url, aws_connect_params = get_aws_connection_info(self.module, boto3=HAS_BOTO3)
+
+        if region:
+            try:
+                self.connection = boto3_conn(self.module, conn_type='client', resource='s3', region=region,
+                                             endpoint=ec2_url,
+                                             **aws_connect_params)
+            except ClientError as e:
+                self.module.fail_json(msg=e.message, exception=traceback.format_exc(),
+                                      **camel_dict_to_snake_dict(e.response))
+        else:
+            self.module.fail_json(msg="AWS region must be specified (like: us-east-1)")
+
+    def get_bucket_list(self):
+        """
+        Return result of list_buckets json encoded
+        :return:
+        """
+        try:
+            buckets = self.connection.list_buckets()
+        except ClientError as e:
+            self.module.fail_json(msg=e.message, exception=traceback.format_exc(),
+                                  **camel_dict_to_snake_dict(e.response))
+
+        return dumps(buckets, cls=DateTimeEncoder)
```

Comment:
What are you wanting the dates to look like?
```
...
        "s3": "{\"Owner\": {\"DisplayName\": \"name\", \"ID\": \"e205714f7bc0d11a268063913d8db524b28b0c6ebac247b2a687650d41\"}, \"Buckets\": [
... {\"CreationDate\": \"2017-06-26T17:49:58+00:00\", \"Name\": \"shertel_test_policy\"}], \"ResponseMetadata\": {\"HTTPStatusCode\": 200, \"RetryAttempts\": 0, \"HostId\": \"O0eBIwO8w/hMqpNGjE04kVsWkF3LS0cElzuFH3F2qUPZvlEbvFqlI8+wISA1nLgah7cx77E=\", \"RequestId\": \"1504C7C1366A8015\", \"HTTPHeaders\": {\"x-amz-id-2\": \"O0eBIwO8w/hMqpNGjE04kVsWkF3LS0cElzuFH3F2qUPZvlEbvFqlI8+wISA1nLgah7cx77E=\", \"server\": \"AmazonS3\", \"transfer-encoding\": \"chunked\", \"x-amz-request-id\": \"1504C7C1366A8015\", \"date\": \"Thu, 27 Jul 2017 12:25:04 GMT\", \"content-type\": \"application/xml\"}}}"
...
```
looks messy to use.
Just returning buckets here without dumps() and DateTimeEncoder provides:
```
        ...
        "s3": {
            "Buckets": [
                 ...
                {
                    "CreationDate": "2017-06-26T17:49:58+00:00",
                    "Name": "shertel_test_policy"
                }
            ],
            "Owner": {
                "DisplayName": "name",
                "ID": "e205714f7bc0d11a268063913d8db524b28b0c6ebac247b2a687650d41"
            },
            "ResponseMetadata": {
                "HTTPHeaders": {
                    "content-type": "application/xml",
                    "date": "Thu, 27 Jul 2017 12:27:04 GMT",
                    "server": "AmazonS3",
                    "transfer-encoding": "chunked",
                    "x-amz-id-2": "npIW2R6ilqlMkGUNlggoqvKSkonkKQoxqBqhOJP3EZ1ZQZubyw2RT6Z3VmT446tcHGg=",
                    "x-amz-request-id": "33E7FD51916D468F"
                },
                "HTTPStatusCode": 200,
                "HostId": "npIW2R6ilqlMkGUNlggoqvKSkonkKQoxqBqhOJP3EZ1ZQZubyw2RT6Z3VmT446tcHGg=",
                "RequestId": "33E7FD51916D468F",
                "RetryAttempts": 0
            }
        }
        ...
```